rclpy/rclpy/executors.py

```diff
@@ -621,10 +621,11 @@ def _wait_for_ready_callbacks(
                     for wt in node.waitables:
                         # Only check waitables that were added to the wait set
                         if wt in waitables and wt.is_ready(wait_set):
-                            handler = self._make_handler(
-                                wt, node, lambda e: (e.take_data(), ), self._execute_waitable)
-                            yielded_work = True
-                            yield handler, wt, node
+                            if wt.callback_group.can_execute(wt):
```

Comment:
I think that this check is just missing.

Reply:
Yes, it definitely missed the check.
I just wondered why https://github.com/ros2/rclpy/blob/013dc2009c9871345317085a9a10504f1df252b2/rclpy/rclpy/executors.py#L619-L628 should be processed before https://github.com/ros2/rclpy/blob/013dc2009c9871345317085a9a10504f1df252b2/rclpy/rclpy/executors.py#L631.
The `waitset` defined at https://github.com/ros2/rclpy/blob/013dc2009c9871345317085a9a10504f1df252b2/rclpy/rclpy/executors.py#L526 can be accessed after https://github.com/ros2/rclpy/blob/013dc2009c9871345317085a9a10504f1df252b2/rclpy/rclpy/executors.py#L631.

Reply:
I tested with the following patch, and it seems to work well with current test cases in `rclpy`.

<details><summary> patch </summary>
<p>

```patch
diff --git a/rclpy/rclpy/executors.py b/rclpy/rclpy/executors.py
index 24fb380..fc67d0d 100644
--- a/rclpy/rclpy/executors.py
+++ b/rclpy/rclpy/executors.py
@@ -616,16 +616,6 @@ class Executor:
                     if gc.handle.pointer in guards_ready:
                         gc._executor_triggered = True
 
-                # Check waitables before wait set is destroyed
-                for node in nodes_to_use:
-                    for wt in node.waitables:
-                        # Only check waitables that were added to the wait set
-                        if wt in waitables and wt.is_ready(wait_set):
-                            handler = self._make_handler(
-                                wt, node, lambda e: (e.take_data(), ), self._execute_waitable)
-                            yielded_work = True
-                            yield handler, wt, node
-
             # Process ready entities one node at a time
             for node in nodes_to_use:
                 for tmr in node.timers:
@@ -671,6 +661,15 @@ class Executor:
                             yielded_work = True
                             yield handler, srv, node
 
+                for wt in node.waitables:
+                    # Only check waitables that were added to the wait set
+                    if wt in waitables and wt.is_ready(wait_set):
+                        if wt.callback_group.can_execute(wt):
+                            handler = self._make_handler(
+                                wt, node, lambda e: (e.take_data(), ), self._execute_waitable)
+                            yielded_work = True
+                            yield handler, wt, node
+
             # Check timeout timer
             if (
                 timeout_nsec == 0 or
```

</p>
</details>

What do you think? Maybe adding a new PR for it could be better.

Reply:
the patch looks okay to work, but behavior is different for application? current code takes out `Waitables` in 1st place if any of them are available, but with your patch that could be after other entities. probably performance improvement?

basically your patch looks good to me, at least `# Check waitables before wait set is destroyed` does not make sense. we probably can issue another PR against this one to have discussion and review.